// styx/lib/estreeFactory.js
var estree_1 = require("./estree");
function createAssignmentExpression(_a)
{
    var left = _a.left, right = _a.right;
    return {
        type: estree_1.NodeType.AssignmentExpression,
        operator: "=",
        left: left,
        right: right
    };
}
exports.createAssignmentExpression = createAssignmentExpression;
function createCallExpression(callee, args)
{
    if (args === void 0) {
        args = []; }
    return {
        type: estree_1.NodeType.CallExpression,
        callee: callee,
        arguments: args
    };
}
exports.createCallExpression = createCallExpression;
function createIdentifier(name)
{
    return {
        type: estree_1.NodeType.Identifier,
        name: name
    };
}
exports.createIdentifier = createIdentifier;
function createIdentityComparisonExpression(_a)
{
    var left = _a.left, right = _a.right;
    return {
        type: estree_1.NodeType.BinaryExpression,
        operator: "===",
        left: left,
        right: right
    };
}
exports.createIdentityComparisonExpression = createIdentityComparisonExpression;
function createLiteral(value)
{
    return {
        type: estree_1.NodeType.Literal,
        raw: String(value),
        value: value
    };
}
exports.createLiteral = createLiteral;
function createUnaryNegationExpression(innerExpression)
{
    return {
        type: estree_1.NodeType.UnaryExpression,
        operator: "!",
        prefix: true,
        argument: innerExpression
    };
}
exports.createUnaryNegationExpression = createUnaryNegationExpression;
