--- /old/Include_.php	2017-06-19 13:54:35.166897827 +0200
+++ ./lib/PHPCfg/Op/Expr/Include_.php	2017-06-19 11:18:03.752808410 +0200
@@ -20,6 +20,8 @@
 
     public $type;
     public $expr;
+	
+    public $included;
 
     public function __construct(Operand $expr, $type, array $attributes = []) {
         parent::__construct($attributes);
@@ -30,4 +32,9 @@
     public function getVariableNames() {
         return ["expr", "result"];
     }
+	
+
+    public function getSubBlocks() {
+        return ["included"];
+    }
 }
